esrally/track/track.py

```diff
@@ -144,6 +144,20 @@ def __eq__(self, other):
         return self.name == other.name
 
 
+class ComponentTemplate(IndexTemplate):
```

Comment:
Do we rely on polymorphism anywhere? I don't think we should rely on inheritance just to safe some code. Also [index templates are deprecated in Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates-v1.html) and we should therefore decouple the two concepts as well in Rally.

Reply:
So whilst i agree we don't seem to, `IndexTemplate` is being used for Composable templates as well - so when we remove support for the deprecated version the class `IndexTemplate` will remain. Consider also how we use the old param source https://github.com/elastic/rally/pull/1099/files/42de57f76c6f103b2b3d20c298684ba76f521bd3#r512749427.

We can still decouple if we're generally against use of polymorphism in rally.

Reply:
Maybe we create a new class `ComposableTemplate`. Both this and `ComponentTemplate` can inherit from `IndexTemplate` for now and later we move `IndexTemplate` to be a ABC? Alternatively we create these two new classes and define in full, leaving a note for them to exploit an ABC once the name `IndexTemplate` is available?

Reply:
I did not imply to be against polymorphism, quite the contrary: We should rely on inheritance if we have polymorphic code. However, I was arguing against using inheritance "just" to save code. The current implementation of `ComponentTemplate` calls the `IndexTemplate` constructor as follows:

```python
super().__init__(name, None, content, delete_matching_indices=False)
```

By specifying less constructor parameters than the base class (the second parameter `pattern` is set to `None`) we indicate that this property is not needed for the subclass, which violates the contract of the base class: every index template requires a matching index pattern. But now that we derive component templates from index templates, which don't require an index pattern, we violate this invariant. Thus, `ComponentTemplate` cannot be a subclass of `IndexTemplate`.

Reply:
Ok i now see your point. I'll pull ComponentTemplate into a separate cluster. When legacy Index template support is removed we should move `IndexTemplate` to an abstract base class with an appropriate contract for both `ComponentTemplate` and `ComposableTemplate`. I'll add a note accordingly.